com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -23,29 +32,119 @@
 	private Set<String> identifierMap = new HashSet<>();
+	private boolean isInitialized = false;
+	private Notifier[] notifiers;
+	private List<Consumer<AdvancedViatraQueryEngine>> initializationActions = new ArrayList<>();
 
-	private ViatraQueryAdapter(AdvancedViatraQueryEngine engine, Project project) {
-		this.engine = engine;
+	private ViatraQueryAdapter(Project project, Notifier[] notifiers) {
 		this.project = project;
+		this.notifiers = notifiers;
 	}
-	
+
+	/**
+	 * @return initialized engine
+	 * @throws ViatraQueryException if engine cannot be initialized yet
+	 * @deprecated Use {@link #getInitializedEngineChecked()} instead
+	 */
+	@Deprecated
 	public AdvancedViatraQueryEngine getEngine() {
+		return getInitializedEngineChecked();
+	}
+
+	/**
+	 * Try to initialize the VIATRA engine if it hasn!t been before and if the 
+	 * initialization is not possible, it throws a ViatraQueryException.
+	 * 
+	 * @return the initialized engine
+	 * @throws ViatraQueryException if initialized engine is not available yet
+	 */
+	public AdvancedViatraQueryEngine getInitializedEngineChecked() {
+		return getInitializedEngine()
+				.orElseThrow(() ->
+					new ViatraQueryException(MESSAGE_ENGINE_NOT_READY, MESSAGE_ENGINE_NOT_READY));
+	}
+
+	/**
+	 * Provide the initialized VIATRA if possible.<br />
+	 * If the engine hasn't been initialized but it can be, then it will be initialize and
+	 * every initialization action registered before will be executed.
+	 * 
+	 * @return the initialized VIATRA engine if possible, otherwise an empty Optional
+	 * @see ViatraQueryAdapter#getInitializedEngine(Consumer)
+	 * @see ViatraQueryAdapter#initializeQueryWhenEngineReady(IQuerySpecification)
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine() {
+		if(!isInitialized) {
+			initializeEngine();
+		}
 		return engine;
 	}
 	
+	/**
+	 * Checks if an engine is available and execute the given action on it if yes.
+	 * If not, the action will be registered in the list of actions which need to be done 
+	 * when the engine is ready and will be executed when the engine become available.
+	 * 
+	 * @param action the operation which would like to use the initialized VIATRA engine
+	 * @return the initialized VIATRA engine of it is available, otherwise empty Optional.
+	 */
+	public Optional<AdvancedViatraQueryEngine> getInitializedEngine(Consumer<AdvancedViatraQueryEngine> action) {
+		if(action != null) {
+			initializationActions.add(action);
+		}
+		return getInitializedEngine();
+	}
+	
+	private Optional<AdvancedViatraQueryEngine> initializeEngine() {
+		if(!isInitialized && ProjectUtilities.isLoaded(project.getPrimaryProject())) {
+			engine = Optional.of(createQueryEngine(project, notifiers));
+			notifiers = new Notifier[0];
+			isInitialized = true;
+		}
+		if(!isInitialized) {
+			LOGGER.warn(MESSAGE_ENGINE_NOT_READY);
+		} else {
+			initializationActions.forEach(action -> action.accept(engine.get()));
```

Comment:
We should wrap the initialization actions into a coalesceTraversal block otherwise it is possible that multiple model traversals might be triggered for the delayed initialization actions. See [VIATRA documentation](https://www.eclipse.org/viatra/documentation/query-api.html#_query_groups_and_coalescing_model_traversals) for details if necessary.

Reply:
I tried to use `coalesceTraversals` but I'm not sure if I used it in the correct way. Please, check the new implementation again.

Reply:
The `coalesceTraversals` is used correctly, but I have opened extra discussions with further issues.